crt_portal/cts_forms/forms.py

```diff
@@ -563,8 +577,50 @@ def __init__(self, *args, **kwargs):
         ]
 
 
+def date_cleaner(self, cleaned_data):
+    day = cleaned_data.get('last_incident_day') or 1
+
+    if day > 31 or day < 1:
+        self.add_error('last_incident_day', ValidationError(
+            _('Please enter a valid day of the month. Day must be between 1 and the last day of the month.')
+        ))
+
+    try:
+        year = cleaned_data['last_incident_year']
+        month = cleaned_data['last_incident_month']
+        test_date = datetime(year, month, day)
+        if test_date > datetime.now():
+            self.add_error('last_incident_year', ValidationError(
+                _('Date can not be in the future.'),
+                params={'value': test_date.strftime('%x')},
+            ))
+        if year < 100:
+            self.add_error('last_incident_year', ValidationError(
+                _('Please enter four digits for the year.'),
+                params={'value': test_date.strftime('%x')},
+            ))
+        if test_date < datetime(1899, 12, 31):
+            self.add_error('last_incident_year', ValidationError(
+                _('Please enter a year after 1900.'),
+                params={'value': test_date.strftime('%x')},
+            ))
+
+    except ValueError:
+        # a bit of a catch-all for all the ways people could make bad dates
+        self.add_error('last_incident_year', ValidationError(
+            _(f'Invalid date format {month}/{day}/{year}.'),
+            params={'value': f'{month}/{day}/{year}'},
+        ))
+    except KeyError:
+        # these will be caught by the built in error validation
+        return cleaned_data
+
+    return cleaned_data
```

Comment:
It _may_ be easier to handle and test the `day` and `year` checks as field level clean functions e.g. `clean_last_incident_year` and leave the ones that _must_ look across fields to this function which is called by `clean` in a few places.

Reply:
IIRC, the day and the year validation from the model won't get called because the date will fail in this clean method first and then it will give the generic bad date message. So we add the day, month, year messages here so that the most specific message will appear.